lib/db_connection/connection_pool.ex

```diff
@@ -0,0 +1,428 @@
+defmodule DBConnection.ConnectionPool do
+
+  @behaviour DBConnection.Pool
+  use GenServer
+  alias DBConnection.ConnectionPool.PoolSupervisor
+
+  @timeout 5000
+  @queue true
+  @queue_target 50
+  @queue_interval 1000
+  @idle_interval 1000
+  @time_unit 1000
+  @holder_key :__info__
+
+  ## DBConnection.Pool API
+
+  @doc false
+  def ensure_all_started(_opts, _type) do
+    {:ok, []}
+  end
+
+  @doc false
+  def start_link(mod, opts) do
+    GenServer.start_link(__MODULE__, {mod, opts}, start_opts(opts))
+  end
+
+  @doc false
+  def child_spec(mod, opts, child_opts \\ []) do
+    Supervisor.Spec.worker(__MODULE__, [mod, opts], child_opts)
+  end
+
+  @doc false
+  def checkout(pool, opts) do
+    queue? = Keyword.get(opts, :queue, @queue)
+    now = System.monotonic_time(@time_unit)
+    timeout = abs_timeout(now, opts)
+    case checkout(pool, queue?, now, timeout) do
+      {:ok, _, _, _} = ok ->
+        ok
+      {:error, _} = error ->
+        error
+      {:exit, reason} ->
+        exit({reason, {__MODULE__, :checkout, [pool, opts]}})
+    end
+  end
+
+  @doc false
+  def checkin({pool, ref, deadline, holder}, conn, _) do
+    cancel_deadline(deadline)
+    now = System.monotonic_time(@time_unit)
+    checkin_holder(holder, pool, conn, {:checkin, ref, now})
+  end
+
+  @doc false
+  def disconnect({pool, ref, deadline, holder}, err, conn, _) do
+    cancel_deadline(deadline)
+    checkin_holder(holder, pool, conn, {:disconnect, ref, err})
+  end
+
+  @doc false
+  def stop({pool, ref, deadline, holder}, err, conn, _) do
+    cancel_deadline(deadline)
+    checkin_holder(holder, pool, conn, {:stop, ref, err})
+  end
+
+  ## Holder api
+
+  @doc false
+  def update(pool, ref, mod, state) do
+    holder = start_holder(pool, ref, mod, state)
+    now = System.monotonic_time(@time_unit)
+    checkin_holder(holder, pool, state, {:checkin, ref, now})
+    holder
+  end
+
+  ## GenServer api
+    
+  def init({mod, opts}) do
+    queue = :ets.new(__MODULE__.Queue, [:private, :ordered_set])
+    {:ok, _} = PoolSupervisor.start_pool(queue, mod, opts)
+    target = Keyword.get(opts, :queue_target, @queue_target)
+    interval = Keyword.get(opts, :queue_interval, @queue_interval)
+    idle_interval = Keyword.get(opts, :idle_interval, @idle_interval)
+    now = System.monotonic_time(@time_unit)
+    codel = %{target: target, interval: interval, delay: 0, slow: false,
+              next: now, poll: nil, idle_interval: idle_interval, idle: nil}
+    codel = start_idle(now, now, start_poll(now, now, codel))
+    {:ok, {:busy, queue, codel}}
+  end
+
+  def handle_info({:db_connection, from, {:checkout, now, queue?}}, {:busy, queue, _} = busy) do
+    case queue? do
+      true ->
+        :ets.insert(queue, {{now, System.unique_integer(), from}})
```

Comment:
What if instead of unique integers we would hold something like a logical clock in the server? Wouldn't it be slightly less overhead?

Reply:
Why would it be less overhead?  I think the code is simpler without changing the server's state.

Reply:
I'm not sure how expensive the `unique_integer()` call is - that's all

Reply:
The zero arity one is very efficient so we should be good. Perhaps if it produces bignums it might make other parts slower. {monotonic_time, unique_integer} is the preferred way to create a unique timestamp, we can move this to the caller if it proves to be cheaper.